src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,138 @@
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new NullPointerException("List can't be null");
+        }
+        int newSize = size + list.size();
+        ensureCapacity(newSize);
 
+        for (int i = 0; i < list.size(); i++) {
+            elements[size + i] = list.get(i);
+        }
+        size = newSize;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkBoundInclusive(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkBoundInclusive(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkBoundInclusive(index);
+        final T removedElement = (T) elements[index];
+        for (int i = index; i < size - 1; i++) {
+            elements[i] = elements[i + 1];
+        }
+        elements[size - 1] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            for (int i = 0; i < size; i++) {
+                if (elements[i] == null) {
+                    final T removedElement = (T) elements[i];
+                    for (int j = i; j < size - 1; j++) {
+                        elements[j] = elements[j + 1];
+                    }
+                    elements[size - 1] = null;
+                    size--;
+                    return removedElement;
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (elements[i] != null && elements[i].equals(element)) {
+                    final T removedElement = (T) elements[i];
+                    for (int j = i; j < size - 1; j++) {
+                        elements[j] = elements[j + 1];
+                    }
+                    elements[size - 1] = null;
+                    size--;
+                    return removedElement;
+                }
+            }
+        }
+        throw new NoSuchElementException("No such element found " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
     }
+
+    public void ensureCapacity(int minCapacity) {
+        if (minCapacity > elements.length) {
+            grow(minCapacity);
+        }
+    }
+
+    public void grow(int minCapacity) {
+        int oldCapacity = elements.length;
+        int currentCapacity = oldCapacity;
+        while (currentCapacity - minCapacity < 0) {
+            currentCapacity = (currentCapacity / HALF_DIVIDER_INDEX) + currentCapacity;
+        }
+        Object[] copyOfElements = new Object[currentCapacity];
+        System.arraycopy(elements, 0, copyOfElements, 0, oldCapacity);
```

Comment:
you don't need copy, you can pass same array twice as src and dest

Reply:
If I understand you correctly, I do not need to create a new array with an increased size, but use an existing one and pass it as arguments to the System.arraycopy method - but then the size remains fixed and I get an [ArrayIndexOutOfBoundsException], can you clarify what you mean please